src/sentry/tagstore/v2/backend.py

```diff
@@ -356,43 +427,63 @@ def incr_group_tag_key_values_seen(self, project_id, group_id, environment_id, k
+        for env in [environment_id, None]:
+            tagkey, _ = self.get_or_create_tag_key(project_id, env, key)
+            tagvalue, _ = self.get_or_create_tag_value(project_id, env, key, value)
+
+            buffer.incr(GroupTagValue,
+                        columns={
+                            'times_seen': count,
+                        },
+                        filters={
+                            'project_id': project_id,
+                            'group_id': group_id,
+                            'environment_id': env,
+                            '_key_id': tagkey.id,
+                            '_value_id': tagvalue.id,
+                        },
+                        extra=extra)
 
     def get_group_event_ids(self, project_id, group_id, environment_id, tags):
-        tagkeys = dict(
-            TagKey.objects.filter(
-                project_id=project_id,
-                key__in=tags.keys(),
-                status=TagKeyStatus.VISIBLE,
-                **self._get_environment_filter(environment_id)
-            ).values_list('key', 'id')
-        )
+        # NOTE: `environment_id=None` needs to be filtered differently in this method.
+        # EventTag never has NULL `environment_id` fields (individual Events always have an environment),
+        # and so `environment_id=None` needs to query EventTag for *all* environments (except, ironically
+        # the aggregate environment, which is NULL).
 
-        tagvalues = {
-            (t[1], t[2]): t[0]
-            for t in TagValue.objects.filter(
-                reduce(or_, (Q(_key__key=k, value=v)
-                             for k, v in six.iteritems(tags))),
-                project_id=project_id,
-                **self._get_environment_filter(environment_id)
-            ).values_list('id', '_key__key', 'value')
-        }
+        if environment_id is None:
+            # filter for all 'real' environments
+            env_filter = {'environment_id__isnull': False}
+        else:
+            env_filter = {'environment_id': environment_id}
+
+        tagkey_qs = TagKey.objects.filter(
+            project_id=project_id,
+            key__in=tags.keys(),
+            status=TagKeyStatus.VISIBLE,
+            **env_filter
+        ).values_list('key', 'id')
+
+        tagkeys = defaultdict(list)
+        for key, key_id in tagkey_qs:
+            tagkeys[key].append(key_id)
+        tagkeys = dict(tagkeys)
+
+        tagvalue_qs = TagValue.objects.filter(
+            reduce(or_, (Q(_key__key=k, value=v)
+                         for k, v in six.iteritems(tags))),
+            project_id=project_id,
+            **env_filter
+        ).values_list('id', '_key__key', 'value')
+
+        tagvalues = defaultdict(list)
+        for value_id, key, value in tagvalue_qs:
+            tagvalues[(key, value)].append(value_id)
+        tagvalues = dict(tagvalues)
 
         try:
             tag_lookups = [(tagkeys[k], tagvalues[(k, v)])
```

Comment:
I don't think the comment below this line is valid any longer?

Reply:
Fixed.